numba/_typeof.c

```diff
@@ -794,8 +795,18 @@ typeof_typecode(PyObject *dispatcher, PyObject *val)
         return typecode_arrayscalar(dispatcher, val);
     }
     /* Array handling */
+    else if (tyobj == &PyArray_Type) {
+         return typecode_ndarray(dispatcher, (PyArrayObject*)val);
+     }
+    /* Subtypes of Array handling */
     else if (PyType_IsSubtype(tyobj, &PyArray_Type)) {
-        return typecode_ndarray(dispatcher, (PyArrayObject*)val);
+        /* If the class has an attribute named __numba_no_subtype_ndarray then
+           don't treat it as a normal variant of a Numpy ndarray but as it's own
+           separate type. */
+        no_subtype_attr = PyObject_HasAttrString(val, "__numba_no_subtype_ndarray__");
+        if (!no_subtype_attr) {
+            return typecode_ndarray(dispatcher, (PyArrayObject*)val);
+        }
```

Comment:
Hmm... do we really have no other way but to add the `__numba_no_subtype_ndarray__`? For now, it does seem like the only way.

Also, we will need test and example for this.

Reply:
For a lot of these places, there were questions in my mind about the right and most consistent way to do it with respect to the rest of the codebase.   See my user code example.